### PR TITLE
Api Updates

### PR DESCRIPTION
**Pull request by JoshLove-msft**
- Adding a new top level client, ServiceBusClient
  - Connection sharing is done implicitly through the top level client
  - ServiceBusConnection is now internal
  - lower level clients are retrieved by calling GetSender/GetReceiver/etc methods on top level client, and specifying the entity name. 
  - top level client has ctors that take connection string, but entity names are ignored, since we require the entity name to be specified when getting lower level client. We don't throw if the entity is included because we want to also allow entity connection strings that may have access restricted to a specific entity. However, operations would fail if you have a connection scoped to a specific entity and you call GetReceiver on a different entity that you don't have permission for.
- Make the ServiceBusConnection be agnostic to the entity name, to allow the same connection to be used for multiple entities.
- Include ManagementLink in both AmqpSender/AmqpReceiver, and refactor operations that were using the ManagementLink from the AmqpClient.
- Other API review updates 
  - make PrefetchCount int rather than uint
  - make ServiceBusMessage ctor take ReadOnlyMemory<byte> rather than byte[]
  - Remove public ctor from ServiceBusReceivedMessage and add static factory method
  - Include Receiver rather than Session object in ProcessMessageEventArgs
  - Rename ServiceBusSession/ServiceBusSubscription properties/types to 
SessionManager/SubscriptionManager
  - Remove ProcessingOptions and instead put the properties directly on the ServiceBusProcessor.
  - Remove ExceptionReceivedContext and pull properties up to ProcessErrorEventArgs.

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpError.cs [hunk 1]
```diff
@@ -216,7 +216,7 @@ internal static class AmqpError
                 if (NotFoundExpression.IsMatch(description)
                     || (description.IndexOf(NotFoundStatusText, StringComparison.InvariantCultureIgnoreCase) >= 0))
                 {
-                    return new ServiceBusException(eventHubsResource, description, ServiceBusException.FailureReason.ResourceNotFound);
+                    return new ServiceBusException(eventHubsResource, description, ServiceBusException.FailureReason.MessageNotFound);
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpError.cs]: nit: eventHubsResource -> serviceBusResource

Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpError.cs]: Fixed!

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs [hunk 2]
```diff
@@ -336,6 +345,179 @@ private void CloseLink(ReceivingAmqpLink link)
...
+        /// <param name="sessionId"></param>
+        /// <param name="propertiesToModify"></param>
+        /// <param name="deadLetterReason"></param>
+        /// <param name="deadLetterDescription"></param>
+        internal override async Task DisposeMessageRequestResponseAsync(
+            Guid[] lockTokens,
+            TimeSpan timeout,
+            DispositionStatus dispositionStatus,
+            bool isSessionReceiver,
+            string sessionId = null,
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs]: Now that we `_isSessionReceiver` and `SessionId` properties in this class, can we remove these two parameters?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs [hunk 3]
```diff
@@ -388,6 +571,47 @@ internal List<ArraySegment<byte>> ConvertLockTokensToDeliveryTags(IEnumerable<Gu
             return lockedUntilUtc;
         }
 
+        /// <summary>
+        ///
+        /// </summary>
+        public override async Task<DateTime> RenewSessionLockAsync(string sessionId, CancellationToken cancellationToken = default)
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs]: Can we remove the `sessionId` param and use the class property?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs [hunk 4]
```diff
@@ -396,7 +620,7 @@ internal List<ArraySegment<byte>> ConvertLockTokensToDeliveryTags(IEnumerable<Gu
         ///
         /// <param name="sessionId"></param>
         /// <param name="timeout"></param>
-        public override async Task<DateTime> RenewSessionLockAsync(
+        internal async Task<DateTime> RenewSessionLockInternal(
             string sessionId,
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs]: Can we remove the `sessionId` param and use the class property?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs [hunk 5]
```diff
@@ -81,240 +81,132 @@ public class ServiceBusReceiverClient : IAsyncDisposable
...
-        ///   If you have defined a shared access policy directly on the Service Bus entity itself, then copying the connection string from that
-        ///   Service Bus entity will result in a connection string that contains the name.
-        /// </remarks>
-        ///
-        public ServiceBusReceiverClient(string connectionString)
-            : this(new ServiceBusConnection(connectionString), new ServiceBusReceiverClientOptions())
+        private static string GetSubscriptionPath(string topicName, string subscriptionName)
         {
-            OwnsConnection = true;
+            return $"{topicName}/{subscriptionName}";
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: We have a same method in `EntityNameFormatter` class. Can we remove this one from here?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs [hunk 6]
```diff
@@ -742,10 +627,14 @@ public virtual async Task<ServiceBusReceivedMessage> PeekAsync(CancellationToken
...
             if (lockTokenGuids.Any(lockToken => RequestResponseLockedMessages.Contains(lockToken)))
             {
-                string receiveLinkName = Consumer.GetReceiveLinkName();
-                return Connection.DisposeMessageRequestResponseAsync(lockTokenGuids, timeout, DispositionStatus.Completed, IsSessionReceiver, Session.UserSpecifiedSessionId, receiveLinkName);
+                return _innerReceiver.DisposeMessageRequestResponseAsync(
+                    lockTokenGuids,
+                    timeout,
+                    DispositionStatus.Completed,
+                    IsSessionReceiver,
+                    SessionManager.SessionId);
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: If we decide to use the `AmqpReceiver` class properties then we can remove IsSessionReceiver and sessionId.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSessionManager.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        {
+            await Task.Delay(1).ConfigureAwait(false);
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        public virtual async Task RenewSessionLockAsync(CancellationToken cancellationToken = default) =>
+            LockedUntilUtc = await _receiver.RenewSessionLockAsync(SessionId, cancellationToken).ConfigureAwait(false);
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSessionManager.cs]: `SessionId` param can be removed, if we decide to use `AmqpReceiver` class property.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs [hunk 8]
```diff
@@ -55,6 +57,7 @@ internal class AmqpConsumer : TransportConsumer
         ///
         private string PartitionId { get; }
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs]: We are not using `partitionId` in Service Bus. Can we remove this?

**Review by ShivangiReja — APPROVED** [review]

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/ManagementUtilities.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.Amqp;
+
+namespace Azure.Messaging.ServiceBus.Amqp
+{
+    /// <summary>
+    /// Utility methods related to the management link
+    /// </summary>
+    public static class ManagementUtilities
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/ManagementUtilities.cs]: Yep

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpError.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpError.cs]: Fixed

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpSender.cs [hunk 10]
```diff
@@ -271,6 +286,201 @@ public override async Task CloseAsync(CancellationToken cancellationToken)
...
+                stopWatch.Stop();
+
+                AmqpResponseMessage amqpResponseMessage = AmqpResponseMessage.CreateResponse(response);
+
+                if (amqpResponseMessage.StatusCode == AmqpResponseStatusCode.OK)
+                {
+                    var sequenceNumbers = amqpResponseMessage.GetValue<long[]>(ManagementConstants.Properties.SequenceNumbers);
+                    if (sequenceNumbers == null || sequenceNumbers.Length < 1)
+                    {
+                        throw new ServiceBusException(true, "Could not schedule message successfully.");
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpSender.cs]: Will cover as part of Exception issue

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs]: Yep!

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs]: Yes

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs]: You bet

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs (hunk 5 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: Yep yep

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs (hunk 6 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: YEs

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSessionManager.cs (hunk 7 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSessionManager.cs]: Removed

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs (hunk 8 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs]: Removed

**Pull request merged**